MalomAPI_cpp/perfect_api.cpp

```diff
@@ -0,0 +1,219 @@
+    }
+    if (s.futureStoneCount(B) > Rules::maxKSZ) {
+        throw std::invalid_argument("Number of stones in blackBitboard + "
+                                    "blackStonesToPlace > " +
+                                    std::to_string(Rules::maxKSZ));
+    }
+
+    std::string errorMsg = s.setOverAndCheckValidSetup();
+    if (errorMsg != "") {
+        throw std::invalid_argument(errorMsg);
+    }
+    if (s.over) {
+        throw std::invalid_argument("Game is already over.");
+    }
+
+    s.lastIrrev = 0;
+
+    int ret = 0;
+
+    try {
+        ret = pp->chooseRandom(pp->goodMoves(s)).toBitBoard();
+    } catch (std::out_of_range &) {
+        throw std::runtime_error("We don't have a database entry for this "
+                                 "position. This can happen either if the "
+                                 "database is corrupted (missing files), or "
+                                 "sometimes when the position is not reachable "
+                                 "from the starting position.");
+    }
+
+    deinitializeIfNeeded();
+
+    return ret;
+}
+
+int MalomSolutionAccess::getBestMoveNoException(
+    int whiteBitboard, int blackBitboard, int whiteStonesToPlace,
+    int blackStonesToPlace, int playerToMove, bool onlyStoneTaking)
+{
+    try {
+        lastError = nullptr;
+        return getBestMove(whiteBitboard, blackBitboard, whiteStonesToPlace,
+                           blackStonesToPlace, playerToMove, onlyStoneTaking);
+    } catch (std::exception &e) {
+        lastError = &e;
+        return 0;
+    }
+}
+
+std::string MalomSolutionAccess::getLastError()
+{
+    if (lastError == nullptr) {
+        return "No error";
+    }
+    return lastError->what();
+}
+
+void MalomSolutionAccess::initializeIfNeeded()
+{
+    if (pp != nullptr) {
+        return;
+    }
+
+    if (sec_val_path == "") {
+        sec_val_path  = ".";
+    }
+
+    Rules::initRules();
+    setVariantStripped();
+    if (!Sectors::hasDatabase()) {
+        throw std::runtime_error("Database files not found in the current "
+                                 "working directory (" +
+                                 std::filesystem::current_path().string() +
+                                 ")");
+    }
+    pp = new PerfectPlayer();
+}
+
+void MalomSolutionAccess::deinitializeIfNeeded()
+{
+    Rules::cleanup();
+
+    if (pp == nullptr) {
+        return;
+    }
+
+    delete pp;
+
+    pp = nullptr;
+}
+
+void MalomSolutionAccess::mustBeBetween(std::string paramName, int value,
+                                        int min, int max)
+{
+    if (value < min || value > max) {
+        throw std::out_of_range(paramName + " must be between " +
+                                std::to_string(min) + " and " +
+                                std::to_string(max));
+    }
+}
+
+void MalomSolutionAccess::setVariantStripped()
+{
+    // copy-paste from Rules.cpp, but references to Main stripped
```

Comment:
We recommend deleting this comment, as it's not really true for the C++ version. (There is no `Rules.cpp`, it's not really a copy-paste in the C++ version, and the other function also doesn't have references to Main in the C++ version.)

Reply:
Fixed in https://github.com/calcitem/Sanmill/commit/7ad0cc5eaa62c142c23c89ea8837ca4304517506